base/messages.go

```diff
@@ -0,0 +1,211 @@
+package base
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"github.com/perlin-network/noise/log"
+	"github.com/perlin-network/noise/protocol"
+	"github.com/pkg/errors"
+	"io"
+	"net"
+)
+
+var _ protocol.MessageAdapter = (*MessageAdapter)(nil)
+
+type MessageAdapter struct {
+	conn              net.Conn
+	finalizerNotifier chan struct{}
+	metadata          map[string]string
+}
+
+func NewMessageAdapterPassive(connAdapter protocol.ConnectionAdapter, conn net.Conn, local []byte, localAddr string) (*MessageAdapter, error) {
+
+	if len(local) > 255 {
+		return nil, errors.New("local or remote id too long")
+	}
+	byteBuf := make([]byte, 1)
+
+	remote := make([]byte, len(local))
+
+	_, err := io.ReadFull(conn, remote)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	_, err = conn.Write(local)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	_, err = io.ReadFull(conn, byteBuf)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	localLen := int(byteBuf[0])
+	if localLen > 0 {
+		localBytes := make([]byte, localLen)
+		_, err = io.ReadFull(conn, localBytes)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+		localAddr = string(localBytes)
+	}
+
+	_, err = io.ReadFull(conn, byteBuf)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	var remoteAddr string
+	remoteLen := int(byteBuf[0])
+	if remoteLen > 0 {
+		remoteBytes := make([]byte, remoteLen)
+		_, err = io.ReadFull(conn, remoteBytes)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+		remoteAddr = string(remoteBytes)
+	}
+
+	adapter := &MessageAdapter{
+		conn:              conn,
+		finalizerNotifier: make(chan struct{}),
+		metadata: map[string]string{
+			"local":      string(local),
+			"remote":     string(remote),
+			"localAddr":  localAddr,
+			"remoteAddr": remoteAddr,
+		},
+	}
+
+	return adapter, nil
+}
+
+func NewMessageAdapterActive(connAdapter protocol.ConnectionAdapter, conn net.Conn, local, remote []byte, localAddr, remoteAddr string) (*MessageAdapter, error) {
+
+	if len(local) > 255 || len(remote) > 255 {
+		return nil, errors.New("local or remote id too long")
+	}
+
+	_, err := conn.Write(local)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+	recvRemote := make([]byte, len(local))
+	_, err = io.ReadFull(conn, recvRemote)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+	if !bytes.Equal(recvRemote, remote) {
+		conn.Close()
+		return nil, errors.Errorf("inconsistent remotes %s and %s", hex.EncodeToString(recvRemote), hex.EncodeToString(remote))
+	}
+
+	if len(remoteAddr) > 255 {
```

Comment:
this and the length check on line 127 can be done after the local/remote length check